### PR TITLE
Add Support to create and upload Debian Packages

### LINKED ISSUE #95: Debian package support
As requested in https://github.com/sass/dart-sass/issues/1164 it would be nice for Dart-SASS and other users (such as [FVM](https://github.com/leoafarias/fvm/issues/343) of this package to have support for producing Debian packages.

If no one wants to tackle this first, I'll have a go at adding support for this when I get a chance.

### PR DESCRIPTION
**Pull request by shubhanshu02**
Fixes #95.

- Import of GPG keys is now available with an option to use external tools to import keys. (Also useful if keys are already present)
- The signing of the release files is now automated.

## **Work Left**
- Documentation for steps to use
- Tests for the module

If you do have any suggestions or corrections, please drop them below. Thanks.

## Description

The feature works by:
1. Creating a .deb package from the given executables. (mentioned in sass/dart-sass#1164)
2. Updating the Release files in the cloned PPA repo and then signing them.
3. Updating the remote PPA repo with the new packages.
4. It creates the Debian packages for both architectures `x64` and `ia32`.

## Usage 

To use this feature, 
1. Set the `debianRepo`, `signingEmail`, `controlData` config variables. 
2. Add the location of control file to `controlData` config.
3. Setup the GPG key import action. The key fingerprint and passphrase are taken from the secrets.
4. Use the `addDebianTasks` function to add the Debian task to Grinder.

### CODE REVIEW

**Comment by stof** [comment]
Making the process automated is a must-have for the sass project, as released are done by GitHub Actions.

File: lib/src/debian.dart [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+Future<String> _createPackageDirectory(String repo, String packageName) async {
+  String debianDir = p.join(repo, packageName);
+  await Directory('$debianDir/DEBIAN').create(recursive: true);
+  await Directory('$debianDir/usr/local/bin').create(recursive: true);
+  return debianDir;
+}
+
+/// Generate the executable files from the map [executables] which contains the
+/// executable name as key and the path to the corresponding Dart file as value.
+void _generateExecutableFiles(String debianDir) {
```
Comment (stof) [inline lib/src/debian.dart:107]: should this task regenerate the executable, or should it use one generated by other release steps @nex3 ?

**Comment by stof** [comment]
Note that the expected usage is that the PPA repo gets exposed using github pages.

**Comment by shubhanshu02** [comment]
> Making the process automated is a must-have for the sass project, as released are done by GitHub Actions.

Yes, I'm working on that. The changes will probably only be in `_releaseNewPackage()` :thinking: 

> Note that the expected usage is that the PPA repo gets exposed using GitHub pages.

Yes, but that has to be enabled manually from the repo settings, right?

File: lib/src/debian.dart [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+      workingDirectory: repo);
+
+  var prefix = "refs/remotes/origin/";
+  if (!result.startsWith(prefix)) {
+    fail('Unexpected output from "git symbolic-ref refs/remotes/origin/HEAD":\n'
+        'Expected a string starting with "$prefix", got:\n'
+        '$result');
+  }
+  return result;
+}
```
Comment (shubhanshu02) [inline lib/src/debian.dart:216]: This function does the same as the function `_originHead()` in `homebrew.md`, but that is a private function (prefixed with `_`). Shall I reuse that function directly in this module?

File: lib/src/debian.dart (hunk 1 above)
Comment (nex3) [reply lib/src/debian.dart:107]: It should depend on an executable-generation task, similar to how the GitHub task does: https://github.com/google/dart_cli_pkg/blob/017d28b68fb9f4305216bfcd41d33ea6afdb708b/lib/src/github.dart#L260-L264

File: lib/src/debian.dart [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        "--default-key",
+        signingEmail.value,
+        "-abs",
+        "-o",
+        "-",
+        "Release",
+      ],
+      workingDirectory: repo);
+  writeString(p.join(repo, 'Release.gpg'), output);
+}
```
Comment (nex3) [inline lib/src/debian.dart:159]: I don't completely understand how this signing process works. Is there any verification that the package author actually owns the email in question? Does it use a private key at all? And if not, what does the signature actually prove?

File: lib/src/debian.dart (hunk 2 above)
Comment (nex3) [reply lib/src/debian.dart:216]: Feel free to move any shared functions into `utils.dart`.

File: lib/src/debian.dart (hunk 3 above)
Comment (stof) [reply lib/src/debian.dart:159]: GPG keys are tied to an email address.

`apt` verifies the signature of APT repositories. So the public key will need to be published, and getting it is part of the instructions to add the apt repository (see for instance the Linux instructions about [installing the Dart SDK](https://dart.dev/get-dart) which first download the public key and pass it to `apt-key add` and then add the APT repository).
The private GPG key is used by this `gpg` command to perform the signing

File: lib/src/debian.dart (hunk 1 above)
Comment (stof) [reply lib/src/debian.dart:107]: well, for apt, it should depend on the linux generation task specifically.

File: lib/src/debian.dart (hunk 2 above)
Comment (shubhanshu02) [reply lib/src/debian.dart:216]: Sure, thanks.

File: lib/src/debian.dart (hunk 3 above)
Comment (shubhanshu02) [reply lib/src/debian.dart:159]: Also, we can export the private and public keys from `gpg` to use on a remote computer (GitHub Actions, in our case) using:

```sh
gpg --output mygpgkey_pub.gpg --armor --export <keyId> # For public key
gpg --output mygpgkey_sec.gpg --armor --export-secret-key <keyId> # For private key
```

And import these to `gpg` using:

```sh
gpg --import ./mygpgkey_pub.gpg
gpg --allow-secret-key-import --import ./mygpgkey_sec.gpg
```

Ref: https://www.debuntu.org/how-to-importexport-gpg-key-pair/

Comment (stof) [reply lib/src/debian.dart:159]: There is https://github.com/marketplace/actions/import-gpg implementing the import of GPG keys for github actions (with proper cleanup to remove the key at the end of the workflow run)

Comment (nex3) [reply lib/src/debian.dart:159]: So if I'm understanding correctly, GPG has some sort of database on the system that it queries using the email as a key? That's a lot to ask an unfamiliar user to set up correctly. Can we provide an optional `aptPrivateKey` config variable that allows users to directly set the private key as a GH Actions secret?

Comment (stof) [reply lib/src/debian.dart:159]: @nex3 I'd rather rely on the dedicated action to import the GPG key (which only requires setting the GPG private key and its passphrase as secret and exposes some output based on the extracted id), as that action also hooks into the cleanup phase of GA to remove the imported key again, avoiding to leak it to the runner.

File: lib/src/debian.dart [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+/// The GitHub repository slug (for example, `username/repo`) of the PPA
+/// repository for this package.
+///
+/// This must be set explicitly.
+final debianRepo = InternalConfigVariable.fn<String>(
+    () => fail("pkg.debianRepo must be set to deploy to PPA repository."));
+
+/// The email address for signing the package.
+final signingEmail = InternalConfigVariable.fn<String>(
```
Comment (stof) [inline lib/src/debian.dart:23]: I suggest using a different name than `email` for the config variable of the GPG key id. The recommended way to reference a private key is to use the fingerprint rather than the email (and the import-gpg action for GHA is providing the fingerprint, so that would be what gets used in the sass automation)

File: lib/src/debian.dart (hunk 3 above)
Comment (stof) [reply lib/src/debian.dart:159]: The workflow would look like that:

```yaml
    steps:
      -
        name: Checkout
        uses: actions/checkout@v2
      -
        name: Import GPG key
        id: import_gpg
        uses: crazy-max/ghaction-import-gpg@v4
        with:
          gpg_private_key: ${{ secrets.GPG_PRIVATE_KEY }}
          passphrase: ${{ secrets.PASSPHRASE }}
      -
        name: Release APT
        run: dart run grinder publish-apt
        env:
          GPG_SIGNING_KEY: ${{ steps.import_gpg.outputs.fingerprint }}
```

File: lib/src/debian.dart (hunk 4 above)
Comment (shubhanshu02) [reply lib/src/debian.dart:23]: Yes, that'd be better. I'll update

File: lib/src/debian.dart (hunk 3 above)
Comment (shubhanshu02) [reply lib/src/debian.dart:159]: If I understood correctly, the main reason for using the actions here is to avoid leakage of the key? @stof

Alternatively, we can import the key from the private key and passphrase, sign the release, and cleanup keys, right from this package. But that's essentially what the action already does :thinking:.

Comment (stof) [reply lib/src/debian.dart]: I vote for letting the action handle all that, which is less code to maintain (and makes the dart_cli_pkg implementation compatible with systems where the GPG key management would have different needs due to the way the GPG key itself is managed)

Comment (stof) [reply lib/src/debian.dart]: note however that I'm not a maintainer here (or in sass), so @nex3 is the one having the final call.

Comment (shubhanshu02) [reply lib/src/debian.dart]: Sorry, I wasn't well for the last few days so wasn't quite active on GitHub. I am on the neutral side here.

##### Reason:
- Less maintenance and no worries about cleanup (pro)
- An extra step that needs to be added manually (con)

What do you say, @nex3?

Comment (nex3) [reply lib/src/debian.dart]: I'd like to support the full signing workflow directly within this package. We can still provide the ability to get the key from the ambient GPG context, which can then be configured using `crazy-max/ghaction-import-gpg`, but one of the goals of this package is to be "batteries-included" and require as little setup as possible.

Comment (stof) [reply lib/src/debian.dart]: Well, the thing is that the current setup still relies on the ambient GPG context (AFAIK, it is not possible to run a self-contained GPG signing command that provides the key directly instead of using the trust store). So this would still depend on having the GPG command available and modifying that ambient context.

Comment (shubhanshu02) [reply lib/src/debian.dart]: If I understood correctly, @nex3, do you mean to say the package shall support two modes as follows?
- use the key from context, i.e., already present in the system. This can also be used if we run the package locally or on a computer where the key is already loaded. Or even in this case, the GitHub action was used to configure GPG.
- load key and then sign using it. (pretty much the release can be made without any external steps)

And @stof, you are saying - even if we code for importing the key, we'll still be using the same GPG commands and do the same stuff like that in the GitHub action?

Comment (nex3) [reply lib/src/debian.dart]: > Well, the thing is that the current setup still relies on the ambient GPG context (AFAIK, it is not possible to run a self-contained GPG signing command that provides the key directly instead of using the trust store). So this would still depend on having the GPG command available and modifying that ambient context.

It's not modifying the ambient context that bothers me, it's the DevX overhead of setting it up.

> If I understood correctly, @nex3, do you mean to say the package shall support two modes as follows?
> 
> * use the key from context, i.e., already present in the system. This can also be used if we run the package locally or on a computer where the key is already loaded. Or even in this case, the GitHub action was used to configure GPG.
> * load key and then sign using it. (pretty much the release can be made without any external steps)

That's right. Basically, we can provide a `pkg.gpgPrivateKey` config variable and, if that's not set, fall back on `pkg.gpgFingerprint` or `pkg.gpgEmail` to look up an existing key from the store.

**Comment by shubhanshu02** [comment]
@nex3, I have added the functions to import the GPG keys. 

Currently, the package imports the control file as a `ConfigVariable<String>`, so users can directly put their control file in the repository. The package uses the same by updating the version (using regex matching). Shall I update it to `ConfigVariable<Map<String, Object>>` or keep it as it is?

File: lib/src/debian.dart [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+
+/// Scan for new .deb packages in the [repo] and update the `Packages` file.
+Future<void> _updatePackagesFile(String repo) async {
+  // Scan for new packages
+  String output = run("dpkg-scanpackages",
+      arguments: ["--multiversion", "."], workingDirectory: repo);
+  // Write the stdout to the file
+  writeString(p.join(repo, 'Packages'), output);
+  // Force Compress the Packages file
+  run("gzip", arguments: ["-k", "-f", "Packages"], workingDirectory: repo);
```
Comment (nex3) [inline lib/src/debian.dart:178]: We should use the [`archive`](https://pub.dev/packages/archive) package for this rather than relying on an external executable. `GZipEncoder().encode(utf8.encode(output))` should work.

File: lib/src/debian.dart [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+  writeString(p.join(repo, 'Release'), output);
+}
+
+/// Sign the Release file with the GPG key.
+Future<void> _updateReleaseGPGFile(String repo) async {
+  run("gpg",
+      arguments: [
+        for (String arg in gpgArgs) arg,
+        "--default-key",
+        (gpgFingerprint.value),
```
Comment (nex3) [inline lib/src/debian.dart:194]: Nit: unnecessary parens here and below

File: lib/src/debian.dart [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+  String output = run("apt-ftparchive",
+      arguments: ["release", "."], workingDirectory: repo);
+  writeString(p.join(repo, 'Release'), output);
+}
+
+/// Sign the Release file with the GPG key.
+Future<void> _updateReleaseGPGFile(String repo) async {
+  run("gpg",
+      arguments: [
+        for (String arg in gpgArgs) arg,
```
Comment (nex3) [inline lib/src/debian.dart:192]:
```suggestion
        ...gpgArgs,
```

File: lib/src/debian.dart [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+        "--default-key",
+        (gpgFingerprint.value),
+        "--passphrase",
+        (gpgPassphrase.value),
+        "-abs",
```
Comment (nex3) [inline lib/src/debian.dart:197]: Short flags are nice for handwritten command-line invocations, but for something in a script file it's clearer to use the full flag names.

File: lib/src/debian.dart [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+void _importGPGPrivateKey() {
+  log("Importing the GPG Private Key");
+  writeString("._privateKey", gpgPrivateKey.value);
+  run(
+    "gpg",
+    arguments: [
+      for (String arg in gpgArgs) arg,
+      "--passphrase",
+      gpgPassphrase.value,
+      "--allow-secret-key-import",
```
Comment (nex3) [inline lib/src/debian.dart:234]: According to the gpg manpage, "this is an obsolete option and is not used anywhere".

File: lib/src/debian.dart [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+        "Release",
+      ],
+      quiet: true,
+      workingDirectory: repo);
+}
+
+/// Import the private key into the GPG.
+void _importGPGPrivateKey() {
+  log("Importing the GPG Private Key");
+  writeString("._privateKey", gpgPrivateKey.value);
```
Comment (nex3) [inline lib/src/debian.dart:227]: Rather than writing this to a temporary file, can you have GPG read it from `/dev/stdin`? You may need to use the lower-level [`Process.start()`](https://api.dart.dev/stable/2.16.0/dart-io/Process/start.html).

**Review by nex3 — CHANGES_REQUESTED** [review]
For the control file, I think the best option is actually to follow the lead of the Homebrew tasks and make the target repository the source-of-truth rather than the local configuration, since that also avoids the risk of overwriting any configuration the user has added there. In other words, rather than exposing `pkg.debianControlData` at all, just assume a control file already exists in the target repo and edit it to have the correct version number.

Also, before this is merged, it'll need tests and docs.

**Comment by nex3** [comment]
I should also mention, thank you for diving into this! I know a bunch of people will appreciate having a Debian package available once this is all set up.

File: lib/src/debian.dart [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+/// repository for this package.
+///
+/// This must be set explicitly.
+final debianRepo = InternalConfigVariable.fn<String>(
+    () => fail("pkg.debianRepo must be set to deploy to PPA repository."));
+
+/// The contents of the debian packages's control file.
+///
+/// By default, it is loaded from `control` at the root of the repository.
+/// It's modifiable.
```
[the remark this answers is not in the document]
Comment (shubhanshu02) [reply lib/src/debian.dart:22]: Sure, I'll remove it.

File: lib/src/debian.dart [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+///
+/// `cli_pkg` will automatically update the `version` field when building the package.
+final controlData = InternalConfigVariable.fn<String>(() =>
```
[the remark this answers is not in the document]
Comment (shubhanshu02) [reply lib/src/debian.dart:25]: Got it. Thanks

File: lib/src/debian.dart [hunk 13]
```diff
@@ -0,0 +1,282 @@
...
+/// By default, this comes from the `GPG_PASSPHRASE` environment variable.
+final gpgPassphrase = InternalConfigVariable.fn<String>(() =>
+    Platform.environment["GPG_PASSPHRASE"] ??
+    fail("pkg.gpgPassphrase must be set to deploy to PPA repository."));
+
+/// The private key for the GPG key to use when signing the release.
+///
+/// By default, it is set as empty. To sign the release, you must set this
+/// after exporting the private key. Or, if using external tools to import the key,
+/// you can leave this empty.
```
[the remark this answers is not in the document]
Comment (shubhanshu02) [reply lib/src/debian.dart:48]: Sure, corrected.

File: lib/src/debian.dart (hunk 8 above)
Comment (shubhanshu02) [reply lib/src/debian.dart:197]: Oh, okay

File: lib/src/debian.dart (hunk 9 above)
Comment (shubhanshu02) [reply lib/src/debian.dart:234]: Didn't notice. Thanks for the catch.

File: lib/src/debian.dart (hunk 5 above)
Comment (shubhanshu02) [reply lib/src/debian.dart:178]: Got it. Will update

File: lib/src/debian.dart [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+    File("control").existsSync()
+        ? File("control").readAsStringSync()
+        : fail("pkg.controlData must be set to update Debian package."));
+
+/// The fingerprint for GPG key to use when signing the release.
+///
+/// By default, this comes from the `GPG_FINGERPRINT` environment variable.
+final gpgFingerprint = InternalConfigVariable.fn<String>(() =>
+    Platform.environment["GPG_FINGERPRINT"] ??
+    fail("pkg.gpgPassphrase must be set to deploy to PPA repository."));
```
[the remark this answers is not in the document]
Comment (shubhanshu02) [reply lib/src/debian.dart:35]: We should be able to determine the fingerprint as the GitHub Action also does that, but I couldn't figure out a way for that. Let me try again.

I read in the manual of GPG command the following and further @stof also suggested using fingerprints over emails.
> The best way to specify a key Id is by using the fingerprint.  This avoids any ambiguities in case that there are duplicated key IDs.

File: lib/src/debian.dart (hunk 10 above)
Comment (shubhanshu02) [reply lib/src/debian.dart:227]: Like echoing the private key and pipelining the output to `gpg`?

**Comment by shubhanshu02** [comment]
> For the control file, I think the best option is actually to follow the lead of the Homebrew tasks and make the target repository the source-of-truth rather than the local configuration, since that also avoids the risk of overwriting any configuration the user has added there.  

That sounds better. I'll update.

> Also, before this is merged, it'll need tests and docs.

Yes, I'll do that. Just waiting to finalize the implementation part.

Also, I wanted to ask how do I apply the code review suggestions - add a new commit or fix up the existing commits? Which would you prefer?

File: lib/src/debian.dart (hunk 14 above)
[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/debian.dart:35]: I can imagine reasons the end user might want to choose to use an email lookup, but I'm fine not adding that unless someone actively asks for it.

File: lib/src/debian.dart (hunk 10 above)
Comment (nex3) [reply lib/src/debian.dart:227]: Essentially yes, but you'd be writing to [`Process.stdin`](https://api.dart.dev/stable/2.16.0/dart-io/Process/stdin.html) yourself.

**Comment by nex3** [comment]
> Also, I wanted to ask how do I apply the code review suggestions - add a new commit or fix up the existing commits? Which would you prefer?

Making new commits is fine, I'll squash and merge in the end.

File: lib/src/debian.dart (hunk 10 above)
Comment (shubhanshu02) [reply lib/src/debian.dart:227]: Ok, I'll try.

File: lib/src/debian.dart (hunk 14 above)
[the remark this answers is not in the document]
Comment (shubhanshu02) [reply lib/src/debian.dart:35]: Passing `email` in `pkg.gpgFingerprint` will work as well since it only references the key. The command doesn't explicitly specify that it's a fingerprint.

[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/debian.dart:35]: Maybe it should just be `pkg.gpgKey` then, to match the CLI parameter name.

[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/debian.dart:35]: I guess that's ambiguous with the contents of the private key itself... I guess `pkg.gpgFingerprint` is fine.

**Comment by shubhanshu02** [comment]
A few changes are still to be implemented that I shall do by tomorrow hopefully.

File: lib/src/debian.dart (hunk 10 above)
Comment (shubhanshu02) [reply lib/src/debian.dart]: I tried using `Process.start()`, but seems like it isn't working properly. I checked the following command works fine:
```sh
cat privateKey | gpg --allow-secret-key-import --import
```

The below piece of code never gets the exitCode and is stuck. In my opinion, it is working similar to running the command `gpg --allow-secret-key-import --import`. This command never ends.

```dart
  var process = await Process.start(
    "gpg",
    [
      ..._gpgArgs,
      "--passphrase",
      gpgPassphrase.value,
      "--import",
    ],
  );
  log('Writing STDIN for the process');
  process.stdin.write(gpgPrivateKey.value);
  var exitCode = await process.exitCode;
  log('Got the exit code');
```

Can you tell where am I going wrong?

File: lib/src/debian.dart (hunk 14 above)
[the remark this answers is not in the document]
Comment (shubhanshu02) [reply lib/src/debian.dart]: Sure. We can still mention in the documentation that passing the email here will also work.

[the remark this answers is not in the document]
Comment (shubhanshu02) [reply lib/src/debian.dart]: Regarding determining the fingerprint from the key, I found that passing `--import-options import-show` while importing gives the following output. So we can extract the fingerprint from the second line in the STDOUT.

```
sec#  rsa4096 2022-01-14 [SC]
      FINGERPRINT
uid            NAME <EMAIL>
ssb#  rsa4096 2022-01-14 [E]
```

[the remark this answers is not in the document]
Comment (shubhanshu02) [reply lib/src/debian.dart]: Found a better fit here using `--with-colons` argument. Here the output can be easily parsed and has a format something like this.

```
....other lines....
fpr:::::::::FINGERPRINT:
...other lines....
```

File: lib/src/debian.dart (hunk 10 above)
Comment (nex3) [reply lib/src/debian.dart]: You need to call `process.stdin.close()` after `process.stdin.write()`. Otherwise, you'll never send the end-of-file marker that tells GPG the private key is complete.

Comment (shubhanshu02) [reply lib/src/debian.dart]: Oh, so the `process.stdin.close()` function sends the EOF to the process. Thanks, it worked.

File: lib/src/debian.dart [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+  freezeSharedVariables();
+  debianRepo.freeze();
+  controlData.freeze();
+  gpgFingerprint.freeze();
+  gpgPassphrase.freeze();
+
+  addTask(GrinderTask('pkg-debian-update',
+      taskFunction: () => _update(),
+      description: 'Update the Debian package.',
+      depends: ['pkg-standalone-linux-x64']));
```
[the remark this answers is not in the document]
Comment (shubhanshu02) [reply lib/src/debian.dart:81]: Regarding this, do you mean publishing both arm64 and x32 architecture packages? Or like an option for which architecture to publish?

[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/debian.dart:81]: Publishing both.

[the remark this answers is not in the document]
Comment (shubhanshu02) [reply lib/src/debian.dart:80]: Got it! I dived a bit into this and found that the 32-bit executable archive packs the AOT snapshot along with a dart runtime. The process is pretty much the same for making the package, so can be done with not many changes I guess.

Currently, I am directly using the `<executable_name>.native` in `build` folder to make the 64-bit packages. But if we are doing this, I feel extracting the packages built in the `pkg-standalone-linux-x32` and `pkg-standalone-linux-x64` seems better.

Shall I directly extract the files from the package using `tar` command? Or decode the packages using `GzipDecoder` and then write each file to their respective location?

[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/debian.dart:80]: I don't think it makes sense to tar up the executables only to immediately untar them. I think it's fine to hard-code the specific files you're accessing, since this is part of the same package that's generating them.

[the remark this answers is not in the document]
Comment (shubhanshu02) [reply lib/src/debian.dart:80]: Actually, the task `pkg-standalone-linux-ia32` makes an archive with `tar.gz` extension in the `build` folder which also has the dart runtime. Along with this, there's an aot snapshot in the `build` folder. And we need the contents of the `tar.gz` file in making the Debian package. That's why I was asking.

[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/debian.dart:80]: I'd rather factor out the logic for downloading the Dart executable.

[the remark this answers is not in the document]
Comment (shubhanshu02) [reply lib/src/debian.dart:80]: Thanks. I almost forgot about that. 

I'm currently nearing the completion of this part, but facing a problem. I am using the `writeString()` function to write the rendered template for the executable script as follows:

```dart
for (var name in executables.value.keys) {
  writeString(
      p.join(executablesPath, name),
      renderTemplate("standalone/executable.sh",
          {"name": standaloneName.value, "executable": name}));
}
```

The problem here is the generated file is not executable, and hence the package shows "Permission Denied" as an error. In manual practice, I can use `chmod +x <filename>` to fix this, but it isn't working in the script. How do I make these files executable?

This is also showing the same error.

```dart
await File(p.join(executablesPath, "src", "dart"))
        .writeAsBytes(dartExecutable);
```

[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/debian.dart:80]: Unfortunately Dart doesn't have a native API for setting file permissions, but you can run `chmod` as a child executable using `run("chmod", arguments: ["+x", path])`.

[the remark this answers is not in the document]
Comment (shubhanshu02) [reply lib/src/debian.dart]: Oh, okay. I have added this in the commit.

**Comment by nex3** [comment]
Go ahead and hit the "re-request review" button once you're ready for another full review.

**Comment by shubhanshu02** [comment]
Actually, currently, I have my mid-term examinations going on. I'll resume as soon as these get over. I'll let you know when the documentation and tests are ready.

**Comment by nex3** [comment]
Good luck with midterms!

**Comment by shubhanshu02** [comment]
@nex3 which tests are expected in this module? Like shall I simulate publishing a test repository or test individual units?

**Comment by nex3** [comment]
Just the same sort of tests that are there for the existing architectures—verify that `pkg-github-*-arm64` uploads the appropriate executable.

**Comment by shubhanshu02** [comment]
Got, it. Thanks.

**Comment by shubhanshu02** [comment]
Hello @nex3, I have been inactive for quite a long time. Sorry for the delay. Can you tell if this work is still available to finish?

**Comment by nex3** [comment]
Yeah, I'm still happy to land this if you add tests.

**Comment by shubhanshu02** [comment]
Thanks. I'll try to complete it soon.

**Comment by nex3** [comment]
Closing this out for now. Feel free to let me know if you get time to work on it again.

**Comment by listepo** [comment]
@shubhanshu02 Hello, are you still interested in this functionality?

**Comment by shubhanshu02** [comment]
HI @listepo! Sorry for the late reply. Just logged in to GitHub after a long time. I do not use `dart` anymore and don't have any such requirement for this now. But the only work left in this PR is unit tests and documentation of steps for using the flow.

**Comment by nikeokoronkwo** [comment]
Hey, since @listepo mentioned that there isn't much left on this, let me see if I can proceed.
